add Prow configs for container image promoter

listx opened the pull request:
Uploading this to receive feedback about how to merge these configs into the main ones in `test-infra/prow/{config,plugins}.yaml`.

AFAICS the configs `test-infra/prow/{config,plugins}.yaml` hold the monolithic prow configs for all Prow jobs.

Some open questions:
- how do we make Prow only schedule these jobs in a new cluster with a limited set of permissions, as suggested in https://github.com/kubernetes/k8s.io/issues/157#issuecomment-465755277 ?
- where will the promoter manifest (YAML file with a list of Docker images to be promoted) live? (github.com/kubernetes/k8s.io ?)

dims commented:
/wg k8s-infra

dims commented:
/ok-to-test

listx commented:
@fejta can you take a look to see how we can add these configs to the main prow folder, and also ensure that we only run these jobs in a separate (new?) cluster, as per your comment https://github.com/kubernetes/k8s.io/issues/157#issuecomment-465755277 ? Or like, if we need to set up a new cluster, maybe this is the right place for it as-is (standalone configs, no need to merge into `test-infra/prow/{config,plugins}.yaml` ?

BenTheElder commented on prow/cip/config.yaml:
```diff
@@ -0,0 +1,153 @@
+    base_ref: master
+  spec:
+    containers:
+    - &cip_prod
```
hehe @fejta will fight you for this :^)

listx replied:
Ack. Will remove YAML anchors/references.

BenTheElder commented on prow/cip/config.yaml:
```diff
@@ -0,0 +1,153 @@
+# Basically anything with a "name: ..." and "spec: ..." is a Prow Job. Prow
```
IMO this should go in something like `config/cip/`, our specific configurations are slowly making their way to a heirarchy in `config/`, decoupled from the underlying tools. eventually prow will probably move to it's own repo.

listx replied:
Ack. I guess the main question I have is --- can we create a new Prow cluster that uses these configs today, or should we piggyback into the existing one https://github.com/kubernetes/test-infra/blob/master/prow/config.yaml ? From your comment I'm going to assume the former (new Prow cluster)

BenTheElder replied:
the config you have here is for a full prow cluster. the presets and jobs could be moved into configs in config/jobs/ and use the existing prow, but we'd need to add webhooks for your repo at minimum.

listx replied:
Ah, makes sense. I'll go ahead and break it up into config/jobs. :sweat_smile: 

As for the plugins.yaml in this PR, I would need to move it into the [main one](https://github.com/kubernetes/test-infra/blob/master/prow/plugins.yaml) though, right?

listx commented:
@fejta PTAL. My main concern is how to make sure these periodics/presubmits/postsubmits run in a cluster with limited privileges (secrets!). I think you mentioned before that we can make jobs run in a dedicated node pool, or something to that effect.

listx commented:
Hmm, the pull-test-infra-bazel check will fail because this PR does not have a corresponding testgrid entry. I'd love to get pointers on how to do this... :thinking:

BenTheElder commented:
breifly: see testgrid/config.yaml, testgrid/README.md has some details (can
elaborate more at another time if necessary)

On Fri, Feb 22, 2019 at 10:10 PM Kubernetes Prow Robot <
notifications@github.com> wrote:

> @listx <https://github.com/listx>: The following test *failed* for commit
> cc2ef36
> <https://github.com/kubernetes/test-infra/commit/cc2ef36150f273a72f547af62260a15b7d922f57>,
> say /retest to rerun them:
> Test name Details Rerun command
> pull-test-infra-bazel link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/test-infra/11414/pull-test-infra-bazel/25091> /test
> pull-test-infra-bazel
>
> Full PR test history <https://gubernator.k8s.io/pr/test-infra/11414>. Your
> PR dashboard <https://gubernator.k8s.io/pr/listx>. Please help us cut
> down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/test-infra/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/11414#issuecomment-466620983>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4Bq_bE5UKZaC_yWea4GxKk-z3L0bXnks5vQNs_gaJpZM4bGktg>
> .
>

listx commented:
/test pull-test-infra-verify-codegen

listx commented:
/assign amwat

spiffxp commented on config/jobs/kubernetes/sig-release/cip/OWNERS:
```diff
@@ -0,0 +1,6 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
+- listx
```
won't block for this, but we should de-SPOF you here

@tpepper as a release-engineering subproject owner?
@dims any interest since you've been helping here?
@hh ?

listx replied:
Oh, good point. I didn't know who to add so I added myself as a minimum. :sweat_smile:

hh replied:
I'd love to help!

spiffxp replied:
this list looks good, thanks!

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,132 @@
+# long periods of time.
+- interval: 24h
+  max_concurrency: 1
+  agent: kubernetes
```
this is implicit, please drop

listx replied:
done

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,132 @@
+            k8s-staging-csi/manifest.yaml,/etc/cip-service-account/service-account.json
+        imagePullPolicy: Always
+
+presets:
```
Will you just put the `service-account-creds` volume and volume mount into your three pod specs and not use a preset?

listx replied:
done

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,132 @@
+      containers:
+      - image: gcr.io/cip-demo-staging/cip@sha256:3823b84b54b2e099f4b6e51e4a6225f04c1bf10c606313fb85a3cfff4df8d318
+        args:
+        - /bin/sh
```
Will you please encapsulate this logic into a `something.sh` script in the kubernetes/k8s.io repo? 

Note that this is how prow does things:
https://github.com/kubernetes/test-infra/blob/1ab0846d5d59ec1ace2a548983164ae7eacc088c/config/jobs/kubernetes/test-infra/test-infra-trusted.yaml#L59-L60

The logic for this we then check into the repo: https://github.com/kubernetes/test-infra/blob/master/prow/push.sh

listx replied:
FWIW the logic was encapsulated like this already: https://github.com/kubernetes/k8s.io/pull/203/files

The reason I didn't encapsulate _all_ of the logic (including the arguments such as `k8s-staging-cluster-api/manifest.yaml,/etc/cip-service-account/service-account.json`) is because the logic that determines the location of the service account creds file lives inside the configuration of this prow job itself.

That being said, are you still opposed to calling the script with arguments like this?

fejta replied:
The thing I don't like is the `/bin/sh`. Is there a reason you cannot use:
```yaml
command:
- multirun.sh
args:
- /app/cip-docker-image.binary
- k8s-staging-cluster-api/manifest.yaml
- ...
```

amwat commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,137 @@
+    # TODO: Move the official cip image to a more serious location. But it's not
+    # a huge problem because we pin it by the digest.
+    - image: gcr.io/cip-demo-staging/cip@sha256:96be240bb072cb6b2b998e29daf87e1de4cd18c18d6ea97bec3f0cce487087ed
+      args:
```
I'm guessing this relies on the default entry-point ? 
Otherwise you want something like `command: /bin/sh` and rest as args. 
Also, we can call `multirun.sh` directly? No need for `/bin/sh -c`

Similar everywhere else.

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,137 @@
+      volumes:
+      - name: k8s-gcr-prod-service-account-creds
+        secret:
+          secretName: k8s-gcr-prod-service-account
```
Can you write a presubmit to validate that only `container-image-promoter.yaml` uses the `k8s-gcr-prod-service-account` secret?

See https://github.com/kubernetes/test-infra/blob/914cddb88731823139411680717f681eb6bd6778/config/tests/jobs/jobs_test.go#L311 for a similar check

listx commented:
I just realized while addressing the comments that I was mounting the production secret even for the dry run (presubmit). This is due to a limitation of the promoter's current design. I think I can add a small patch to the promoter so that dry runs of the promoter use 0 service accounts (to guarantee that it won't touch anything during dry runs). Let's wait for that first.

listx commented:
Update: I have another PR for the promoter tool that should be resolved first before we move forward with this PR: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/31

listx commented:
Update: I have another PR for the promoter tool https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/32. Waiting on that before addressing the comments on this PR.

fejta commented on config/jobs/kubernetes/sig-release/cip/OWNERS:
```diff
@@ -0,0 +1,12 @@
+- hh
+- listx
+- tpepper
+reviewers:
```
If these lists are the same (and they should be) can you just have the approvers section?

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -71,6 +71,38 @@ postsubmits:
+        - k8s-staging-cluster-api/manifest.yaml,/etc/k8s-gcr-prod-service-account/service-account.json
+        - k8s-staging-coredns/manifest.yaml,/etc/k8s-gcr-prod-service-account/service-account.json
+        - k8s-staging-csi/manifest.yaml,/etc/k8s-gcr-prod-service-account/service-account.json
+        imagePullPolicy: Always
```
This doesn't seem super necessary if you're pulling an explicit sha as opposed to a dynamic tag

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -71,6 +71,38 @@ postsubmits:
+    cluster: test-infra-trusted
+    decorate: true
+    run_if_changed: ".*/manifest.yaml"
+    skip_report: false
```
Don't we want to leave a note on the commit saying whether the job passed?

listx replied:
I thought that's what `skip_report: false` did already?

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -319,4 +351,60 @@ periodics:
+    volumes:
+    - name: k8s-gcr-prod-service-account-creds
+      secret:
+        secretName: k8s-gcr-prod-service-account
```
I forget, does this secret exist in the service cluster yet?

listx replied:
Not yet --- can we create it after this PR is merged?

fejta replied:
Secret now exists

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -71,6 +71,38 @@ postsubmits:
         secret:
           secretName: pusher-service-account
+  kubernetes/k8s.io:
+  - name: cip-prodrun
```
This should be `post-k8sio-cip`

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,29 @@
+presubmits:
+  kubernetes/k8s.io:
+  - name: cip-dryrun
```
this should be `pull-k8sio-cip`

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -319,4 +351,60 @@ periodics:
+- interval: 24h
+  max_concurrency: 1
+  # This name is the "job name", passed in as "--job=NAME" for mkpj.
+  name: cip-sanity
```
this should be `ci-k8sio-cip`

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,29 @@
+        - k8s-staging-coredns/manifest.yaml
+        - k8s-staging-csi/manifest.yaml
+        env:
+        - name: CIP_OPTS
```
Can invert this so that it runs dry by default and the prod jobs set `--dry-run=false`? This seems like the safer, more appropriate default

fejta commented on config/jobs/kubernetes/sig-release/cip/container-image-promoter.yaml:
```diff
@@ -0,0 +1,26 @@
+    - ^master$
+    spec:
+      containers:
+      - image: gcr.io/cip-demo-staging/cip@sha256:8ce6fba6cc5873648dd8bb8f9c591be0959e07b78a0381c796df2b359f990908
```
It would be nice if there was some way to trace this to a commit. This is why we use names more like `cip-20190328-deadbeef` also if we document where/how to build this image (at least for the trusted variant of this job)

listx replied:
Currently I just build it locally and push, but I do agree that there needs to be some more official way to do things (including naming the image a certain way, promoting it, etc). These concerns will be addressed in the near future.

fejta replied:
I mean just a comment about where the source code to this image lives. If you die, how is someone going to create the next version of the image? 

is it https://github.com/kubernetes-sigs/k8s-container-image-promoter I guess?

fejta commented on config/jobs/kubernetes/test-infra/test-infra-trusted.yaml:
```diff
@@ -319,4 +352,61 @@ periodics:
+    containers:
+    # TODO: Move the official cip image to a more serious location. But it's not
+    # a huge problem because we pin it by the digest.
+    - image: gcr.io/cip-demo-staging/cip@sha256:20190401-v1.0.0-3-g44d6169
```
a) are you sure this is a digtest? Looks weirder than I would expect.

Also I would still like to see instructions on how to review the contents of this image

listx replied:
Oh my, how embarrassing. Thank you so much for catching this... will fix.

listx replied:
Fixed in 66c42a3; PTAL?

listx commented:
/hold cancel
